src/firefly/js/ui/panel/CollapsiblePanel.jsx

```diff
@@ -25,85 +24,64 @@ export const CollapseHeaderCorner = {
                             BottomLeft: 0b1000
                         };
 
-function collapsibleStateFromProps(props) {
-    // component key needs to be defined if the state needs to be saved though unmount/mount
-    let isOpen = props.componentKey && getComponentState(props.componentKey).isOpen;
-    if (!isBoolean(isOpen)) {
-        // use property to initialize state
-        isOpen = props.isOpen? props.isOpen: false;
-    }
 
-    return {isOpen};
-}
 
-export class CollapsiblePanel extends PureComponent {
+export function CollapsiblePanel({ componentKey, header, isOpen=false, headerRoundCorner=0, headerStyle, contentStyle,
+                                     style, borderStyle=CollapseBorder.Oneborder, onToggle, children }) {
+    function getIsOpenFromComponentState() {
+        if (!componentKey) return isOpen;
+        return getComponentState(componentKey)?.isOpen ?? isOpen;
+    }
+    const isOpenGlobalState = getIsOpenFromComponentState();
 
-    constructor(props) {
-        super(props);
+    const [isOpenState, setIsOpenState] = useState(isOpenGlobalState);
 
-        this.state = collapsibleStateFromProps(props);
+    // if global state has updated, update local state
+    if (isOpenGlobalState !== isOpenState) setIsOpenState(isOpenGlobalState);
```

Comment:
This line can't be here. You should not set state in the render. I should only be in a callback or an effect.

```js
useEffect( () => {
   if (isOpenGlobalState !== isOpenState) setIsOpenState(isOpenGlobalState);
},[isOpenState,isOpenGlobalState]
```

Reply:
Well actually I did that earlier but then removed it (see [this diff](https://github.com/Caltech-IPAC/firefly/pull/1285/commits/5426ab1fe20c825e4198e4223404c6ba62b1d912)) because according to [react docs](https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops), `getDerivedStateFromProps` that was doing this job in class component, is in fact setting the state during rendering. I know getDerivedStateFromProps is discouraged but wasn't sure if setting state after rendering is more appropriate.

Also this made me thinking, shouldn't we pass state from parent component to child component as props rather than passing it to global redux state and then checking in child component (with a hook) if a change has occurred in that piece of global state to update local state of child? I'd appreciate if you can point me to our code or redux docs section that explain the best way to implement such data flow.

Reply:
I am glad you pointed that out because that is what has been bothering me the whole time about this component. 

It made to work on two modes, one way is as part of a FieldGroup (`FieldGroupCollapsible`) which effectively gets it state from redux via out field group (see `FieldGroupCntrl.jsx`).

But that is not the problem, it is the other mode which is getting the state from the store _if_ componentKey is defined.

There are two modes but to make this work correctly there should be a view component that only works with props.

So do this.

Make a `CollapsiblePanelView` component takes `isOpen` but not `componentKey`. change `FieldGroupCollapsible` to use it.  Then make a new component small `CollapsiblePanel` component that uses `useStoreConnector` which is a hooks that update when the store updates.  

It would look something like this.
```js
export function CollapsiblePanel({componentKey, isOpen, ...rest }) {

    const isOpenState= useStoreConnector(() => {
        return getComponentState(componentKey)?.isOpen ?? isOpen;
    });

    return (
        <CollapsiblePanelView {...{ ...rest, isOpen:isOpenState,
            onToggle: (newIsOpen) => dispatchComponentStateChange(componentKey, {isOpen: newIsOpen}) }}
        />);

}
```

This way we have  pure view component and two redux connect component that use it.

Reply:
@robyww Thank you for this explanation and suggestion. It took me a while to understand this but I understood it eventually. And data flow in the component is definitely more readable now after splitting out view as pure component.